zkevm-circuits/src/sig_circuit/ecdsa.rs

```diff
@@ -126,57 +130,27 @@ where
         base_chip.limb_bits,
         var_window_bits,
     );
-    let point_at_infinity = EcPoint::construct(
-        ecc_chip
-            .field_chip()
-            .load_constant(ctx, fe_to_biguint(&CF::zero())),
-        ecc_chip
-            .field_chip()
-            .load_constant(ctx, fe_to_biguint(&CF::zero())),
-    );
     let u2_is_zero =
         base_chip
             .range()
             .gate()
             .or(ctx, Existing(s_is_zero), Existing(is_pubkey_zero));
     let u2_mul = ecc_chip.select(ctx, &point_at_infinity, &u2_mul, &u2_is_zero);
-    // compute u3*G this is directly assigned for G so no scalar_multiply is required
-    let u3_mul = {
-        let generator = GA::generator();
-        let neg_generator = -generator;
-        let generator = ecc_chip.assign_constant_point(ctx, generator);
-        let neg_generator = ecc_chip.assign_constant_point(ctx, neg_generator);
-        ecc_chip.select(ctx, &generator, &neg_generator, &u1_is_one)
-    };
-
-    // compute u2 * pubkey + u3 * G
-    base_chip.enforce_less_than_p(ctx, u2_mul.x());
-    base_chip.enforce_less_than_p(ctx, u3_mul.x());
-    let u2_pk_u3_g = ec_add_unequal(base_chip, ctx, &u2_mul, &u3_mul, false);
 
-    // check
-    // - (u1 + u3) * G
-    // - u2 * pubkey + u3 * G
-    // are not equal
-    let u1_u2_x_eq = ecc_chip.is_equal(ctx, &u1u3_mul, &u2_pk_u3_g);
-    let u1_u2_not_eq = base_chip.range.gate().not(ctx, Existing(u1_u2_x_eq));
-
-    // compute (x1, y1) = u1 * G + u2 * pubkey and check (r mod n) == x1 as integers
-    // which is basically u1u3_mul + u2_mul - u3_mul
-    // WARNING: For optimization reasons, does not reduce x1 mod n, which is
-    //          invalid unless p is very close to n in size.
-    //
-    // WARNING: this may be trigger errors if:
-    // - u1u3_mul == u2_mul
-    //
-    // if r is sampled truly from random then this will not happen
-    // to completely ensure the correctness we may need to sample u3 from random, but it is quite
-    // costly.
-    let sum = ec_add_unequal(base_chip, ctx, &u1u3_mul, &u2_mul, false);
+    // =================================
+    // case 2:
+    // =================================
+    // compute tmp = u1 G + u2 Pk + random point and check if tmp == random point
+    let random_point = ecc_chip.load_random_point::<GA>(ctx);
+    let tmp = ecc_chip.add_unequal(ctx, &u1_mul, &random_point, false);
+    let tmp = ecc_chip.add_unequal(ctx, &u2_mul, &tmp, false);
+    let sum_is_infinity = ecc_chip.is_equal(ctx, &tmp, &random_point);
+    let sum_is_not_infinity = base_chip
+        .gate()
+        .not(ctx, QuantumCell::Existing(sum_is_infinity));
+    let sum = ecc_chip.sub_unequal(ctx, &tmp, &random_point, false);
```

Comment:
We should add a test case where `u1.G + u2.PK` is in fact point at infinity.

For such a case, since `tmp == random_point`, the `sub_unequal` could possibly panic?